File: app/src/main/java/com/soundcloud/android/discovery/MultipleContentSelectionCardRenderer.java
package com.soundcloud.android.discovery;

import butterknife.ButterKnife;
import com.soundcloud.android.R;
import com.soundcloud.android.discovery.DiscoveryCard.MultipleContentSelectionCard;
import com.soundcloud.android.model.Urn;
import com.soundcloud.android.presentation.CellRenderer;
import com.soundcloud.java.optional.Optional;

import android.content.Context;
import android.os.Parcelable;
import android.support.annotation.IdRes;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import javax.inject.Inject;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class MultipleContentSelectionCardRenderer implements CellRenderer<MultipleContentSelectionCard> {

    private final Map<Urn, Parcelable> scrollingState = new HashMap<>();
    private final SelectionItemAdapterFactory selectionItemAdapterFactory;

    @Inject
    MultipleContentSelectionCardRenderer(SelectionItemAdapterFactory selectionItemAdapterFactory) {
        this.selectionItemAdapterFactory = selectionItemAdapterFactory;
    }

    @Override
    public View createItemView(ViewGroup viewGroup) {
        View view = LayoutInflater.from(viewGroup.getContext())
                                     .inflate(R.layout.multiple_content_selection_card, viewGroup, false);
        initCarousel(view, ButterKnife.findById(view, R.id.selection_playlists_carousel));
        return view;
    }

    private void initCarousel(View cardView, RecyclerView recyclerView) {
        final Context context = recyclerView.getContext();
        final SelectionItemAdapter adapter = selectionItemAdapterFactory.create();

        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false));
        recyclerView.setAdapter(adapter);

        cardView.setTag(adapter);
    }

    @Override
    public void bindItemView(int position, View view, List<MultipleContentSelectionCard> list) {
        MultipleContentSelectionCard selectionCard = list.get(position);
        bindTitle(view, selectionCard);
        bindDescription(view, selectionCard);
        bindCarousel(view, selectionCard);
    }

    private void bindTitle(View view, MultipleContentSelectionCard selectionCard) {
        bindText(view, R.id.selection_title, selectionCard.title());
    }

    private void bindDescription(View view, MultipleContentSelectionCard selectionCard) {
        bindText(view, R.id.selection_description, selectionCard.description());
    }

    private void bindText(View view, @IdRes int id, Optional<String> text) {
        TextView textView = ButterKnife.findById(view, id);
        if (text.isPresent()) {
            textView.setVisibility(View.VISIBLE);
            textView.setText(text.get());
        } else {
            textView.setVisibility(View.GONE);
        }
    }

    private void bindCarousel(View view, MultipleContentSelectionCard selectionCard) {
        final RecyclerView recyclerView = ButterKnife.findById(view, R.id.selection_playlists_carousel);
        final SelectionItemAdapter adapter = (SelectionItemAdapter) view.getTag();

        saveOldScrollingState(adapter, recyclerView);
        adapter.updateSelection(selectionCard);
        loadScrollingState(adapter, recyclerView);
    }

    private void saveOldScrollingState(SelectionItemAdapter adapter, RecyclerView recyclerView) {
        adapter.selectionUrn().ifPresent(urn -> scrollingState.put(urn, recyclerView.getLayoutManager().onSaveInstanceState()));
    }

    private void loadScrollingState(SelectionItemAdapter adapter, RecyclerView recyclerView) {
        if (adapter.selectionUrn().isPresent() && scrollingState.containsKey(adapter.selectionUrn().get())) {
            recyclerView.getLayoutManager().onRestoreInstanceState(scrollingState.get(adapter.selectionUrn().get()));
        } else {
            recyclerView.scrollToPosition(0);
        }
    }
}

File: app/src/main/java/com/soundcloud/android/discovery/MultipleContentSelectionCardRenderer.kt
package com.soundcloud.android.discovery

import android.os.Parcelable
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.soundcloud.android.R
import com.soundcloud.android.discovery.DiscoveryCardViewModel.MultipleContentSelectionCard
import com.soundcloud.android.model.Urn
import com.soundcloud.android.presentation.CellRenderer
import com.soundcloud.android.utils.OpenForTesting
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject
import kotlinx.android.synthetic.main.discovery_multiple_content_selection_card.view.*
import javax.inject.Inject

@OpenForTesting
class MultipleContentSelectionCardRenderer
@Inject
constructor(private val selectionItemAdapterFactory: SelectionItemAdapter.Factory) : CellRenderer<MultipleContentSelectionCard> {

    private val scrollingState = mutableMapOf<Urn, Parcelable>()
    internal val selectionItemInCardClickListener: PublishSubject<SelectionItemViewModel> = PublishSubject.create()

    internal fun selectionItemClick(): Observable<SelectionItemViewModel> {
        return selectionItemInCardClickListener
    }

    override fun createItemView(viewGroup: ViewGroup): View {
        val view = LayoutInflater.from(viewGroup.context)
                .inflate(R.layout.discovery_multiple_content_selection_card, viewGroup, false)
        initCarousel(view, view.selection_playlists_carousel)
        return view
    }

    private fun initCarousel(cardView: View, recyclerView: RecyclerView) {
        val context = recyclerView.context
        val adapter = selectionItemAdapterFactory.create(selectionItemInCardClickListener)

        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        recyclerView.adapter = adapter

        cardView.tag = adapter
    }

    override fun bindItemView(position: Int, view: View, list: List<MultipleContentSelectionCard>) {
        val selectionCard = list[position]
        bindTitle(view, selectionCard)
        bindDescription(view, selectionCard)
        bindCarousel(view, selectionCard)
    }

    private fun bindTitle(view: View, selectionCard: MultipleContentSelectionCard) {
        bindText(view.selection_title, selectionCard.title)
    }

    private fun bindDescription(view: View, selectionCard: MultipleContentSelectionCard) {
        bindText(view.selection_description, selectionCard.description)
    }

    private fun bindText(textView: TextView, text: String?) {
        if (text != null) {
            textView.visibility = View.VISIBLE
            textView.text = text
        } else {
            textView.visibility = View.GONE
        }
    }

    private fun bindCarousel(view: View, selectionCard: MultipleContentSelectionCard) {
        val recyclerView = view.selection_playlists_carousel

        (view.tag as? SelectionItemAdapter)?.apply {
            saveOldScrollingState(this, recyclerView)
            this.updateSelection(selectionCard)
            loadScrollingState(this, recyclerView)
        }
    }

    private fun saveOldScrollingState(adapter: SelectionItemAdapter, recyclerView: RecyclerView) {
        adapter.selectionUrn?.let { urn -> scrollingState.put(urn, recyclerView.layoutManager.onSaveInstanceState()) }
    }

    private fun loadScrollingState(adapter: SelectionItemAdapter, recyclerView: RecyclerView) {
        val selectionUrn = adapter.selectionUrn
        if (selectionUrn != null && scrollingState.containsKey(selectionUrn)) {
            recyclerView.layoutManager.onRestoreInstanceState(scrollingState[selectionUrn])
        } else {
            recyclerView.scrollToPosition(0)
        }
    }
}
